File: consumer/consumer.py
# consumer/consumer.py
import pika
import json
import os
import time
import sys

RABBIT_HOST = os.getenv("RABBIT_HOST", "rabbitmq")
RABBIT_USER = os.getenv("RABBIT_USER", "user")
RABBIT_PASS = os.getenv("RABBIT_PASS", "password")
EXCHANGE = "demo_exchange"

def connect_with_retries(retries=10, delay=1):
    credentials = pika.PlainCredentials(RABBIT_USER, RABBIT_PASS)
    params = pika.ConnectionParameters(host=RABBIT_HOST, credentials=credentials)
    attempt = 0
    while True:
        try:
            conn = pika.BlockingConnection(params)
            return conn
        except Exception as e:
            attempt += 1
            if attempt >= retries:
                print(f"[!] No se pudo conectar después de {attempt} intentos: {e}", file=sys.stderr)
                raise
            wait = delay * attempt
            print(f"[!] Conexión fallida (intento {attempt}/{retries}), reintentando en {wait}s... ({e})")
            time.sleep(wait)

def on_message(ch, method, properties, body):
    try:
        payload = json.loads(body.decode())
    except Exception:
        payload = body.decode()
    print(f"[v] Mensaje recibido: {payload}", flush=True)
    ch.basic_ack(delivery_tag=method.delivery_tag)

def main():
    connection = connect_with_retries()
    channel = connection.channel()
    channel.exchange_declare(exchange=EXCHANGE, exchange_type='fanout', durable=True)

    q = channel.queue_declare(queue='', exclusive=True)
    queue_name = q.method.queue
    channel.queue_bind(exchange=EXCHANGE, queue=queue_name)

    print("Esperando mensajes. Ctrl+C para salir.")
    channel.basic_consume(queue=queue_name, on_message_callback=on_message)
    try:
        channel.start_consuming()
    except KeyboardInterrupt:
        print("Consumer detenido.")
    finally:
        try:
            connection.close()
        except Exception:
            pass

if __name__ == "__main__":
    main()

File: Escenario1/consumer3/consumer.go
// consumer3/consumer.go
package main

import (
	"fmt"
	"log"
	"os"

	amqp "github.com/rabbitmq/amqp091-go"
)

func getenv(k, d string) string {
	if v := os.Getenv(k); v != "" { return v }
	return d
}

func main() {
	host := getenv("RABBIT_HOST", "rabbitmq")
	user := getenv("RABBIT_USER", "guest")
	pass := getenv("RABBIT_PASS", "guest")
	url := fmt.Sprintf("amqp://%s:%s@%s:5672/", user, pass, host)

	conn, err := amqp.Dial(url)
	if err != nil { log.Fatalf("dial: %v", err) }
	defer conn.Close()

	ch, err := conn.Channel()
	if err != nil { log.Fatalf("chan: %v", err) }
	defer ch.Close()

	ex := "demo_exchange"
	if err := ch.ExchangeDeclare(ex, "fanout", true, false, false, false, nil); err != nil {
		log.Fatalf("exchange declare: %v", err)
	}

	q, err := ch.QueueDeclare("", false, true, true, false, nil) // name="", exclusive -> cola anónima
	if err != nil { log.Fatalf("queue declare: %v", err) }

	if err := ch.QueueBind(q.Name, "", ex, false, nil); err != nil {
		log.Fatalf("queue bind: %v", err)
	}

	msgs, err := ch.Consume(q.Name, "", true, true, false, false, nil)
	if err != nil { log.Fatalf("consume: %v", err) }

	fmt.Println("Go Consumer esperando mensajes...")
	for m := range msgs {
		fmt.Println("[Go] Mensaje recibido:", string(m.Body))
	}
}
